src/extract_gpuinfo_mali_common.c

```diff
@@ -0,0 +1,528 @@
+  if (gpu_info->last_update_process_cache) {
+    struct mali_process_info_cache *cache_entry, *tmp;
+    HASH_ITER(hh, gpu_info->last_update_process_cache, cache_entry, tmp) {
+      HASH_DEL(gpu_info->last_update_process_cache, cache_entry);
+      free(cache_entry);
+    }
+  }
+  gpu_info->last_update_process_cache = gpu_info->current_update_process_cache;
+  gpu_info->current_update_process_cache = NULL;
+}
+
+void mali_common_get_running_processes(struct gpu_info *_gpu_info, enum mali_version version) {
+  // For Mali, we register a fdinfo callback that will fill the gpu_process datastructure of the gpu_info structure
+  // for us. This avoids going through /proc multiple times per update for multiple GPUs.
+  struct gpu_info_mali *gpu_info = container_of(_gpu_info, struct gpu_info_mali, base);
+  if (gpu_info->version != version) {
+    fprintf(stderr, "Wrong device version: %u\n", gpu_info->version);
+    abort();
+  }
+
+  swap_process_cache_for_next_update(gpu_info);
+}
+
+void mali_common_parse_fdinfo_handle_cache(struct gpu_info_mali *gpu_info,
+					   struct gpu_process *process_info,
+					   nvtop_time current_time,
+					   uint64_t total_cycles,
+					   unsigned cid,
+					   bool engine_count)
+{
+  struct mali_process_info_cache *cache_entry;
+  struct unique_cache_id ucid = {.client_id = cid, .pid = process_info->pid};
+
+  HASH_FIND_CLIENT(gpu_info->last_update_process_cache, &ucid, cache_entry);
+
+  if (cache_entry) {
+    uint64_t time_elapsed = nvtop_difftime_u64(cache_entry->last_measurement_tstamp, current_time);
+    SET_GPUINFO_PROCESS(process_info, sample_delta, time_elapsed);
+    if (engine_count)
+      SET_GPUINFO_PROCESS(process_info, gpu_cycles, total_cycles - cache_entry->last_cycles);
+    cache_entry->last_cycles = total_cycles;
+    HASH_DEL(gpu_info->last_update_process_cache, cache_entry);
+    if (GPUINFO_PROCESS_FIELD_VALID(process_info, gfx_engine_used) &&
+        MALI_CACHE_FIELD_VALID(cache_entry, engine_render) &&
+        // In some rare occasions, the gfx engine usage reported by the driver is lowering (might be a driver bug)
+        process_info->gfx_engine_used >= cache_entry->engine_render &&
+        process_info->gfx_engine_used - cache_entry->engine_render <= time_elapsed) {
+      SET_GPUINFO_PROCESS(
+          process_info, gpu_usage,
+          busy_usage_from_time_usage_round(process_info->gfx_engine_used, cache_entry->engine_render, time_elapsed));
+    }
+  } else {
+    cache_entry = calloc(1, sizeof(*cache_entry));
+    if (!cache_entry)
+      return;
+    cache_entry->client_id.client_id = cid;
+    cache_entry->client_id.pid = process_info->pid;
+    cache_entry->last_cycles = total_cycles;
+  }
+
+#ifndef NDEBUG
+  // We should only process one fdinfo entry per client id per update
+  struct mali_process_info_cache *cache_entry_check;
+  HASH_FIND_CLIENT(gpu_info->current_update_process_cache, &cid, cache_entry_check);
```

Comment:
We fixed some uniqueness issue for the id in #248, I'll push a patch to the pull request to apply the same fix.